src/libstd/ffi/c_str.rs

```diff
@@ -1551,6 +1551,20 @@ impl ops::Index<ops::RangeFull> for CString {
     }
 }
 
+#[stable(feature = "cstr_range_from", since = "1.45.0")]
+impl ops::Index<ops::RangeFrom<usize>> for CStr {
+    type Output = CStr;
+
+    #[inline]
+    fn index(&self, index: ops::RangeFrom<usize>) -> &CStr {
+        let bytes = &self.to_bytes()[index.start..];
+        unsafe {
+            let slice = slice::from_raw_parts(bytes.as_ptr(), bytes.len() + 1);
```

Comment:
I think this will make Miri unhappy and that you have to use `self.as_ptr()` and `self.len()` instead.

cc @RalfJung

Reply:
Indeed -- `bytes` and all pointers derived from it may only be used to access the memory inside the `bytes` slice, which is violated by the `len+1` here. Good catch!

Reply:
not sure how I'm supposed to use `self.as_ptr()` and `self.len()` here. the purpose of `&self.to_bytes()[index.start..]` is to slice off the left part of the `CStr`, which means both the pointer and length will be different from those of `self`.

`bytes` derives from `.to_bytes()` (which leaves off the null byte at the end), which derives from `.to_bytes_with_nul()`,  and that's what I try to recover by doing `bytes.len() + 1`. the length should still be valid because it accounts for the null byte that was left off.

Reply:
`bytes` does not contain the NULL byte. It is incorrect to construct from that a slice that goes beyond the scope of `bytes`, i.e., to re-add the NULL byte.

Reply:
okay. if i use `bytes.as_ptr()` to calculate an offset from `self.to_bytes_with_nul()`, would that be okay? something like this:

```rust
let bytes = &self.to_bytes()[index.start..];
let bytes_with_nul = self.to_bytes_with_nul();
unsafe {
    let offset = bytes.as_ptr().offset_from(bytes_with_nul.as_ptr()) as usize;
    CStr::from_bytes_with_nul_unchecked(&bytes_with_nul[offset..])
}
```
seems it would need a `#![feature(ptr_offset_from)]` tho, is that okay?

Reply:
Why do all that when you could just do the intended comparison between range start and string length directly and, if that extra check passes, do what the original implementation did (slice the &[u8] and turn it back into a CStr)?

Reply:
i guess i just wanted to do the clever thing and reuse the panic message from the slice directly, but you're right, i should just do the simpler thing.